InvalidUndefBytes: Track size of undef region used

divergentdave opened the pull request:
This PR adds a size to `UndefinedBehaviorInfo::InvalidUndefBytes`, to keep track of how many undefined bytes in a row were accessed, and changes a few methods to pass this information through. This additional information will eventually be used in Miri to improve diagnostics for this UB error. See also rust-lang/miri#1354 for prior discussion.

I expect Miri will break the next time its submodule is updated, due to this change to the `InvalidUndefBytes`. (The current commit for src/tools/miri predates rust-lang/miri#1354, and thus does not try to destructure the `InvalidUndefBytes` variant) I have a corresponding change prepared for that repository.

r? @RalfJung

oli-obk commented on src/librustc_middle/mir/interpret/error.rs:
```diff
@@ -358,7 +358,10 @@ pub enum UndefinedBehaviorInfo {
     /// Using uninitialized data where it is not allowed.
-    InvalidUndefBytes(Option<Pointer>),
+    InvalidUndefBytes {
+        ptr: Option<Pointer>,
```
Since these are always either both `None` or both `Some`, please create a separate datastructure holding the `Pointer` and the `Size` and use only a single `Option`.

RalfJung replied:
Or maybe just use a tuple? Not sure if it is worth naming that.

oli-obk commented on src/librustc_middle/mir/interpret/value.rs:
```diff
@@ -583,7 +583,7 @@ impl<'tcx, Tag> ScalarMaybeUndef<Tag> {
         match self {
             ScalarMaybeUndef::Scalar(scalar) => Ok(scalar),
-            ScalarMaybeUndef::Undef => throw_ub!(InvalidUndefBytes(None)),
+            ScalarMaybeUndef::Undef => throw_ub!(InvalidUndefBytes { ptr: None, size: None }),
```
we may at some point want to store the size of the undef in the `Undef` just like `Scalar::Raw` has a `size` field.

oli-obk replied:
Though if we do that, we may want to have the `ptr` and `size` `Option`s separate as you have it now ...

RalfJung replied:
Not sure if we do, the point of the `Size` in the error is to point at a memory dump and point out the region of uninitialized memory that was accessed. This here is a by-value piece of uninitialized data, so there is nothing to point at.

oli-obk commented on src/librustc_mir/interpret/validity.rs:
```diff
@@ -782,7 +782,7 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValueVisitor<'mir, 'tcx, M>
                         match err.kind {
-                            err_ub!(InvalidUndefBytes(Some(ptr))) => {
+                            err_ub!(InvalidUndefBytes { ptr: Some(ptr), .. }) => {
                                 // Some byte was uninitialized, determine which
```
We could also generate a range by using the size to detect whether multiple elements are unintialized. Probably not worth the effort though, especially since it's only moderately helpful if there are initialized pieces in between.

RalfJung commented:
One thing I am wondering about in terms of the Miri diagnostics that will be based on this: is "first uninit range" really the most useful piece of information? I could imagine it would be more useful to learn what the full memory range of the access was, and there in that access the uninit stuff is.

@divergentdave you seem to have a good use for these diagnostics as you initially contributed them, what do you think?

bors commented:
:umbrella: The latest upstream changes (presumably #71704) made this pull request unmergeable. Please resolve the merge conflicts.

RalfJung commented:
@divergentdave this PR needs rebasing to resolve conflicts and there are some review comments to take care of. No pressure though, if you need more time that's all right. :)

oli-obk commented on src/librustc_middle/mir/interpret/allocation.rs:
```diff
@@ -545,17 +546,23 @@ impl<'tcx, Tag: Copy, Extra> Allocation<Tag, Extra> {
-    /// at which the first undefined access begins.
-    fn is_defined(&self, ptr: Pointer<Tag>, size: Size) -> Result<(), Size> {
+    /// Returns `Ok(())` if it's defined. Otherwise returns the range of byte
+    /// indexes of the first continguous undefined access.
```
```suggestion
    /// indexes of the first contiguous undefined access.
```

divergentdave replied:
Ack, whoops! Fixed both places and pushed.

oli-obk commented on src/librustc_middle/mir/interpret/allocation.rs:
```diff
@@ -758,19 +765,25 @@ impl InitMask {
-    /// Returns `Ok(())` if it's initialized. Otherwise returns the index of the byte
-    /// at which the first uninitialized access begins.
+    /// Returns `Ok(())` if it's initialized. Otherwise returns a range of byte
+    /// indexes for the first continguous span of the uninitialized access.
```
```suggestion
    /// indexes for the first contiguous span of the uninitialized access.
```

divergentdave commented:
I rebased everything up and changed the InvalidUninitBytes variant to hold an Option of a new struct, which now holds pointers and sizes for both the original memory access and the subset that was uninitialized.

RalfJung commented on src/librustc_middle/mir/interpret/error.rs:
```diff
@@ -327,6 +327,15 @@ impl fmt::Display for CheckInAllocMsg {
+    pub access_ptr: Pointer,
+    pub access_size: Size,
+    pub uninit_ptr: Pointer,
+    pub uninit_size: Size,
```
Please add doc comments to all fields.

RalfJung commented on src/librustc_middle/mir/interpret/error.rs:
```diff
@@ -384,7 +393,7 @@ pub enum UndefinedBehaviorInfo<'tcx> {
     InvalidStr(std::str::Utf8Error),
     /// Using uninitialized data where it is not allowed.
-    InvalidUninitBytes(Option<Pointer>),
+    InvalidUninitBytes(Option<UninitBytesAccess>),
```
This looks like it makes the variant rather big... does this increase the size of `InterpError`?

divergentdave replied:
Yeah, it increased it from 32 bytes to 64 bytes. I'll put it behind a Box to bring it back down.

RalfJung replied:
Could you please add a `static_assert_size!` to the error type?

RalfJung commented on src/librustc_middle/mir/interpret/error.rs:
```diff
@@ -455,10 +464,17 @@ impl fmt::Display for UndefinedBehaviorInfo<'_> {
                 f,
-                "reading uninitialized memory at {}, but this operation requires initialized memory",
-                p
+                "reading {} byte{} of memory at {:?}, but {} byte{} {} uninitialized at {:?}, \
```
I think you can use `{}` to format pointers these days.

RalfJung replied:
Also maybe "starting at" would be clearer than "at"?

RalfJung commented on src/librustc_middle/mir/interpret/error.rs:
```diff
@@ -384,7 +397,7 @@ pub enum UndefinedBehaviorInfo<'tcx> {
     InvalidStr(std::str::Utf8Error),
     /// Using uninitialized data where it is not allowed.
-    InvalidUninitBytes(Option<Pointer>),
+    InvalidUninitBytes(Option<Box<UninitBytesAccess>>),
```
To avoid the size hit, boxing certainly makes sense. But OTOH we have a `allocates` function below to make sure that errors coming up during ConstProp do not allocate as that would be a perf hit. But I think this error here actually can come up during ConstProp so we probably cannot add it there. @oli-obk what do you think?

divergentdave replied:
I updated `allocates()` and the tests still pass. It seems like InvalidUninitBytes(None) from reading a Scalar is far more common, and the Some case comes up with things like FFI with libc. Perhaps that means it doesn't come up in ConstProp?

RalfJung replied:
Hm... that might be a coincidence though...

RalfJung commented on src/librustc_middle/mir/interpret/error.rs:
```diff
@@ -455,10 +468,18 @@ impl fmt::Display for UndefinedBehaviorInfo<'_> {
-                "reading uninitialized memory at {}, but this operation requires initialized memory",
-                p
+                "reading {} byte{} of memory starting at {}, \
+                but {} byte{} {} uninitialized starting at {}, \
+                and this operation requires initialized memory",
```
```suggestion
                 but {} byte{} {} uninitialized starting at {}, \
                 and this operation requires initialized memory",
```

RalfJung commented on src/librustc_middle/mir/interpret/error.rs:
```diff
@@ -327,6 +327,22 @@ impl fmt::Display for CheckInAllocMsg {
+}
+
+#[cfg(target_arch = "x86_64")]
+static_assert_size!(UndefinedBehaviorInfo<'_>, 32);
```
Actually what I meant is doing this for `InterpError`. We don't really care about the individual variants.

divergentdave replied:
Whoops, got it

RalfJung commented on src/librustc_middle/mir/interpret/error.rs:
```diff
@@ -556,6 +577,9 @@ impl dyn MachineStopType {
 }
 
+#[cfg(target_arch = "x86_64")]
+static_assert_size!(InterpError<'_>, 40);
```
Uff, 40 bytes. @oli-obk we might have to work on reducing that before @nnethercote catches us.^^

RalfJung commented:
Looking good, thanks!
@bors r+

bors commented:
:pushpin: Commit 0148a7f26ce636ac22ac7797dcd7d292c59e8576 has been approved by `RalfJung`

<!-- @bors r=RalfJung 0148a7f26ce636ac22ac7797dcd7d292c59e8576 -->
<!-- homu: {"type":"Approved","sha":"0148a7f26ce636ac22ac7797dcd7d292c59e8576","approver":"RalfJung"} -->

bors commented:
:hourglass: Testing commit 0148a7f26ce636ac22ac7797dcd7d292c59e8576 with merge a9ca1ec9280ca1e5020edd699917c3367a30a798...
<!-- homu: {"type":"BuildStarted","head_sha":"0148a7f26ce636ac22ac7797dcd7d292c59e8576","merge_sha":"a9ca1ec9280ca1e5020edd699917c3367a30a798"} -->
